### PR TITLE
Vet clinic database: add "join table" for visits

### PR DESCRIPTION
**Pull request by najibullahjafari**
- I created a table named vets with the following columns:

- [ ] id: integer (set it as autoincremented PRIMARY KEY)
- [ ] name: string
- [ ] age: integer
- [ ] date_of_graduation: date

- I inserted the following data for vets:

- [ ] Vet William Tatcher is 45 years old and graduated Apr 23rd, 2000.
- [ ] Vet Maisy Smith is 26 years old and graduated Jan 17th, 2019.
- [ ] Vet Stephanie Mendez is 64 years old and graduated May 4th, 1981.
- [ ] Vet Jack Harkness is 38 years old and graduated Jun 8th, 2008.
- [ ] Insert the following data for specializations:
- [ ] Vet William Tatcher is specialized in Pokemon.
- [ ] Vet Stephanie Mendez is specialized in Digimon and Pokemon.
- [ ] Vet Jack Harkness is specialized in Digimon.
- [ ] Insert the following data for visits:
- [ ] Agumon visited William Tatcher on May 24th, 2020.
- [ ] Agumon visited Stephanie Mendez on Jul 22th, 2020.
- [ ] Gabumon visited Jack Harkness on Feb 2nd, 2021.
- [ ] Pikachu visited Maisy Smith on Jan 5th, 2020.
- [ ] Pikachu visited Maisy Smith on Mar 8th, 2020.
- [ ] Pikachu visited Maisy Smith on May 14th, 2020.
- [ ] Devimon visited Stephanie Mendez on May 4th, 2021.
- [ ] Charmander visited Jack Harkness on Feb 24th, 2021.
- [ ] Plantmon visited Maisy Smith on Dec 21st, 2019.
- [ ] Plantmon visited William Tatcher on Aug 10th, 2020.
- [ ] Plantmon visited Maisy Smith on Apr 7th, 2021.
- [ ] Squirtle visited Stephanie Mendez on Sep 29th, 2019.
- [ ] Angemon visited Jack Harkness on Oct 3rd, 2020.
- [ ] Angemon visited Jack Harkness on Nov 4th, 2020.
- [ ] Boarmon visited Maisy Smith on Jan 24th, 2019.
- [ ] Boarmon visited Maisy Smith on May 15th, 2019.
- [ ] Boarmon visited Maisy Smith on Feb 27th, 2020.
- [ ] Boarmon visited Maisy Smith on Aug 3rd, 2020.
- [ ] Blossom visited Stephanie Mendez on May 24th, 2020.
- [ ] Blossom visited William Tatcher on Jan 11th, 2021.

- I wrote queries to answer the following:

- [ ] Who was the last animal seen by William Tatcher?
- [ ] How many different animals did Stephanie Mendez see?
- [ ] List all vets and their specialties, including vets with no specialties.
- [ ] List all animals that visited Stephanie Mendez between April 1st and August 30th, 2020.
- [ ] What animal has the most visits to vets?
- [ ] Who was Maisy Smith's first visit?
- [ ] Details for most recent visit: animal information, vet information, and date of visit.
- [ ] How many visits were with a vet that did not specialize in that animal's species?
- [ ] What specialty should Maisy Smith consider getting? Look for the species she gets the most.
Here you can see the screenshots:
Screenshot 2023-09-01 172958
Screenshot 2023-09-01 173915
Screenshot 2023-09-01 174432
Screenshot 2023-09-01 175324
Screenshot 2023-09-01 180905
Screenshot 2023-09-01 181223
Screenshot 2023-09-01 181546
Screenshot 2023-09-01 182230

### CODE REVIEW

File: queries.sql [hunk 1]
```diff
@@ -117,4 +117,84 @@ FROM owners o
...
+
+
+-- Questions 
+-- Who was the last animal seen by William Tatcher?
+SELECT a.name 
+FROM animals a 
+JOIN visits v ON a.id = v.animals_id
+JOIN vets vet ON v.animals_id=vet.id
+WHERE vet.name = 'Vet William Tatcher '
+ORDER BY v.visits_date;
```
Comment (souad988) [inline queries.sql]: The returned value of this query will be a list of animal names that match the specified conditions. The result will be ordered by the visits_date, so you will get the names of animals in ascending order of their visit dates by 'Vet William Tatcher'.  
the question was Who was the last animal seen by William Tatcher? so we need to order by date in descendent order then limit the result to one item.

File: queries.sql [hunk 2]
```diff
@@ -117,4 +117,84 @@ FROM owners o
...
+-- How many different animals did Stephanie Mendez see?
+SELECT COUNT(*)
+FROM vets v 
+JOIN visits vi ON v.id = vi.vets_id
+WHERE v.name = 'Vet Stephanie Mendez ';
```
Comment (souad988) [inline queries.sql]: great you are so close , this query will return total number of visits including redandancies , to correct you can use 
`SELECT COUNT(DISTINCT a.id)
FROM vets v
JOIN visits vi ON v.id = vi.vets_id
JOIN animals a ON vi.animals_id = a.id
WHERE v.name = 'Vet Stephanie Mendez';`

refer to this article:
[https://www.commandprompt.com/education/how-do-i-count-unique-values-in-postgresql/](https://www.commandprompt.com/education/how-do-i-count-unique-values-in-postgresql/)

File: queries.sql [hunk 3]
```diff
@@ -117,4 +117,84 @@ FROM owners o
...
+-- List all vets and their specialties, including vets with no specialties.
+SELECT v.name 
+FROM vets v 
+LEFT JOIN specializations s ON v.id = s.vet_id
+LEFT JOIN species ss ON ss.id = s.species_id
+GROUP BY v.name, v.id;
```
Comment (souad988) [inline queries.sql]: Good but the question here is to return vets data and their specialities and the query you provided returns only name and id , you can consider this solution and also add a group by clause to return each name with their respective specialities
`  SELECT vets.*, species.name
  FROM vets
  LEFT JOIN specializations ON vets.id = specializations.vet_id
  LEFT JOIN species ON specializations.species_id = species.id;
`

**Review by souad988 — CHANGES_REQUESTED** [review]
Hi @najibullahjafari , 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

242722065-db428b0c-8dfc-49c4-a156-42f227e32868


## Highlights
Good file structure
Good readme file

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by Strangeal — APPROVED** [review]
Hi @najibullahjafari, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 




## Highlights
- Add requested changes ✅ 

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**